### ISSUE TITLE
[tabular] What to do with class imbalance + extremely small sample size

### ISSUE THREAD

**Issue by Innixma** [autogluon/autogluon#4791]
## The Problem

When class imbalance is present with extremely small sample sizes, models using early stopping can produce very poor test scores while having very good validation scores.

I'm unsure if there is a way to approach this problem that will get a better result on average outside of not using early stopping and instead using a fixed number of epochs, but I will showcase an example failure case below of mainline AutoGluon.

## Possible Solutions

1. Only fit models that don't use early stopping (RandomForest, ExtraTrees, KNN, zeroshot TabPFN). Use validation for ensembling or pick single best.
2. Same as (1), but get rid of validation and train with everything. Use TabRepo to learn what single model is best on average, only fit that one model. Don't use ensembling.
3. Same as (1), but also include other models that normally do ES, but instead don't use ES and instead fit much fewer epochs/iterations.
4. Do something clever with sample weights.
5. Use bagging, maybe this just fixes a lot of the problems?

## Example failure case

In this scenario, we use a train_data of 10. This contains 8 samples of class 0 and 2 samples of class 1.

2 samples are selected for the validation set, but both of the samples are from class 0.

The longer the model trains, the more overfit it becomes towards predicting 0 very strongly on the two validation samples.

This leads to the model training for 1000 epochs and having an extremely high validation score, but a very bad test score.

At the same time, random forest actually gets a reasonable test score, but its validation score isn't overfit and thus it isn't selected in the final weighted ensemble.

### Code

```python
from autogluon.tabular import TabularPredictor
from autogluon_benchmark.tasks.task_wrapper import OpenMLTaskWrapper


if __name__ == '__main__':
    task = OpenMLTaskWrapper.from_name(dataset="adult")

    train_data, test_data = task.get_train_test_split_combined(
        fold=0,
        train_size=10,
    )
    label = task.label
    problem_type = task.problem_type

    predictor = TabularPredictor(label=label, problem_type=problem_type, eval_metric="log_loss")

    predictor = predictor.fit(
        train_data=train_data,
        hyperparameters={
            "NN_TORCH": [{}],
            "RF": [{}],
        },
    )

    leaderboard = predictor.leaderboard(test_data, display=True)
```

### Leaderboard Output

```
                 model  score_test  score_val eval_metric  pred_time_test  pred_time_val  fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0         RandomForest   -0.512300  -0.491349    log_loss        0.080387       0.039243  2.038273                 0.080387                0.039243           2.038273            1       True          1
1  WeightedEnsemble_L2   -1.901183  -0.008147    log_loss        0.028655       0.009030  3.668050                 0.001320                0.000444           0.005774            2       True          3
2       NeuralNetTorch   -1.901183  -0.008147    log_loss        0.027335       0.008586  3.662276                 0.027335                0.008586           3.662276            1       True          2
```

**Comment by Killer3048** [comment]
@Innixma My version
For extreme class imbalance with tiny datasets in AutoGluon:
- Disable unstable models and force class awareness:
```python
hyperparameters = {
    "RF": {"class_weight": "balanced", "max_depth": 4},
    "XT": {"class_weight": "balanced"},
    "TABPFN": {},
    "KNN": {}
}
```
- Aggressive validation stabilization:
```python
predictor.fit(
    num_bag_folds=2,  # Max folds for n=10
    num_bag_sets=5,   # Repeated stratified splits
    auto_stack=False, # Disable deceptive ensembles
    calibrate=True    # Fix probability outputs
)
```
- Use ROC-AUC metric instead of log loss:
```python
TabularPredictor(..., eval_metric="roc_auc")
```
Eliminates validation overfit via repeated bagging
TabPFN provides free prior knowledge (no training)
Class weighting + shallow trees prevent majority-class bias
ROC-AUC ignores class distribution skew
